Introduce "transformSpec" at ingest-time.

gianm opened the pull request:
It accepts a "filter" (standard query filter object) and "transforms" (a
list of objects with "name" and "expression"). These can be used to do
filtering and single-row transforms without need for a separate data
processing job.

The "expression" fields use the same expression language as other
expression-based feature.

Not yet documented for two reasons:

1. Expression-based features are generally not yet documented (although I think the time is soon arriving where we should document them)
2. There is not really a clear place to put the docs, since dataSchema docs are strewn about the documentation landscape. This is probably a sign that the docs need some refactoring.

I'm hoping to address both of these in a later patch, but for now, the docs are unchanged.

drcrallen commented:
@gianm can you go in a little deeper on the motivation here? 

Previously the druid stuff was solely dedicated to indexing. Is the purpose here to add in more advanced ETL because people don't want to run more than one type of "worker"? or is it because there's common very primitive ETLs that will help 85% of use cases you see?

As an alternative, what if we had more "indexing" extensions that plugged into different data writers or data syncs of different systems. Like a spark writer / hadoop output format. Or stuff specialized for certain other data processing frameworks?

gianm commented:
@drcrallen The motivation is to add some very basic stateless transforms to the Druid indexer to save people from the cost and hassle of running extra jobs when they want to do something basic. It's the same idea as the flattenSpec that exists for JSON and Avro.

Some examples of problems that this solves:

- I have a field called "url" but I only want to index the domain -> use a regexp_extract expression
- I have two fields, "firstName" and "lastName" but I want to index them as a pair -> use a concat expression
- I have two time columns (perhaps start time and settlement time for a transaction), I want one to be the primary timestamp but I want to store the other as an int64 dimension -> use a timestamp_parse expression
- I want to normalize some field to lowercase -> use a lower expression

fjy commented:
👍

gianm commented:
@drcrallen Does that rationale seem reasonable to you?

himanshug commented on server/src/main/java/io/druid/segment/indexing/Transformer.java:
```diff
@@ -0,0 +1,232 @@
+    if (valueMatcher != null) {
+      rowSupplierForValueMatcher.set(transformedRow);
+      if (!valueMatcher.matches()) {
+        return null;
```
most indexers don't have the ability to filter null rows but rather fail with NPE , does this filtering work with either hadoop indexing job or kafka indexing service ?

gianm replied:
@himanshug that's a good point, I tested it with IndexTask but not every form. IndexTask works since it has this code:

```
          // The null inputRow means the caller must skip this row.
          if (inputRow == null) {
            continue;
          }
```

I'll add some tests for kafka, hadoop, and realtime tasks as well.

himanshug commented on server/src/main/java/io/druid/segment/indexing/Transformer.java:
```diff
@@ -0,0 +1,232 @@
+    if (transforms.isEmpty()) {
+      transformedRow = row;
+    } else {
+      transformedRow = new TransformedInputRow(row, transforms);
```
also, can you adjust https://github.com/druid-io/druid/blob/master/indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceRecordReader.java#L145 for hadoop re-indexing where rows don't go through the configured parser.

gianm replied:
Hmm, I think IngestSegmentFirehoseFactory doesn't work right either, it also bypasses the parser. Will try to do something to make this work for both of them and add some tests.

himanshug replied:
yeah its even better to handle this in `IngestSegmentFirehoseFactory` then `DatasourceRecordReader` will get it automatically.

himanshug commented:
@drcrallen @gianm sounds good for a feature, i have also seen the needs for simple transformations at ingestion time.
does undocumented in this case mean that you don't want users to use this feature just yet because things may change very rapidly or that you want to do more testing and ensure it works ?

gianm commented:
@himanshug just undocumented because nothing expression-related is documented yet. I'm intending to do another patch before too long that will add documentation for all of them, linking through to the existing `math-expr.md`.

Also for the reason in the original comment -- there is not really a clear place to put docs for stuff in dataSchema, since the docs probably need a refactor.

gianm commented:
Marking WIP until I can add more tests.

gianm commented:
Tests added, removing WIP tag.

gianm commented:
Removed WIP tag.

- Added tests for index task, realtime task, kafka task, hadoop mapper, and ingestSegment firehose
- Changed Transform to an interface (not just hard-coded to use expressions)

gianm commented:
@himanshug any further comments?

gianm commented:
@drcrallen @himanshug any further comments?

himanshug commented:
i'll get back to this today.

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java:
```diff
@@ -75,10 +76,19 @@ protected void map(
           throw e;
         }
-        log.debug(e, "Ignoring invalid row [%s] due to parsing error", value.toString());
+        if (log.isDebugEnabled()) {
```
log.isDebugEnabled() check is built-in inside log.debug(..) so this is not necessary. you probably wanted to avoid eager `value.toString()` computation that can be changed by replacing same with `value` .

gianm replied:
Yes, this was the reason. Good point, I'll change it to just `value`.

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java:
```diff
@@ -75,10 +76,19 @@ protected void map(
 
+      if (inputRow == null) {
+        // Throw away null rows from the parser.
+        if (log.isDebugEnabled()) {
```
same here

himanshug commented on server/src/main/java/io/druid/segment/indexing/Transform.java:
```diff
@@ -0,0 +1,34 @@
+})
+public interface Transform
+{
+  String getName();
```
do we really need to have this method ?

gianm replied:
Yes, it's like the name of an aggregator, you need to have it so we know what the field should be called. If you want a field `z` that is `x + y` then it would be specified like,

```
"transformSpec" : {
  "transforms" : [
    { "type" : "expression", "name" : "z", "expression" : "x + y" }
  ]
}
```

himanshug replied:
ah, i see, its name of the column in the row ... would be nice to add little doc on this to be clear for other implementers.

gianm replied:
I added javadocs to the interface.

himanshug commented on server/src/main/java/io/druid/segment/indexing/Transformer.java:
```diff
@@ -0,0 +1,202 @@
+      return row.compareTo(o);
+    }
+
+    private Object getValueFromRow(final String column)
```
doesn't appear to be used

himanshug commented on server/src/main/java/io/druid/segment/indexing/TransformSpec.java:
```diff
@@ -0,0 +1,139 @@
+    } else if (parser instanceof TransformingStringInputRowParser) {
+      return ((TransformingStringInputRowParser) parser).getTransformSpec();
+    } else {
+      return TransformSpec.NONE;
```
should we throw a RuntimeException in this case because this should only get called in the codepaths where a `decorate(..)` would've happened. If we reached here that means there is some coding error ?

gianm replied:
This case can happen if the transformSpec is a no-op, in which case `decorate` will return the original parser, not a decorated one.

himanshug replied:
ok, this makes coding errors harder to catch though. if we always did the wrapping (costs only one object) in `decorate` and threw exception here then catching coding errors would be trivial.

but, at this point its more of a nit than anything so feel free to ignore.

gianm replied:
Hmm. I had done it this way so that if you're not using transformSpec, it has no effect (parser isn't wrapped), mostly so just in case there was some bug with decorated parsers it won't affect all users. Although I could have it always wrap the parser so bugs (if any) are found more quickly.

What do you think?

gianm commented:
@himanshug pushed a new commit addressing your comments.

gianm commented:
@himanshug I would say a transform _can_ be used to add columns (as long as those columns are functions of existing ones…). And I guess you could use it to sort of "remove" a column in a sense, by overwriting it with one that is just all nulls? Anyway, I just pushed a new javadoc that is more clear.

I also took your suggestion in https://github.com/druid-io/druid/pull/4890#discussion_r146064560 and added the sanity check.

himanshug commented:
@gianm dint reallize all the possibilties, so that doc helps. thanks.
